src/object/srv_obj_migrate.c

```diff
@@ -810,6 +813,105 @@ migrate_fetch_update_parity(struct migrate_one *mrone, daos_handle_t oh,
 	return rc;
 }
 
+static int
+migrate_fetch_update_single(struct migrate_one *mrone, daos_handle_t oh,
+			    struct ds_cont_child *ds_cont)
+{
+	struct daos_oclass_attr	*oca;
+	d_sg_list_t	 	sgls[DSS_ENUM_UNPACK_MAX_IODS];
+	d_iov_t		 	iov[DSS_ENUM_UNPACK_MAX_IODS] = { 0 };
+	char		 	*data;
+	daos_size_t	 	size;
+	int		 	i;
+	int		 	rc;
+
+	oca = daos_oclass_attr_find(mrone->mo_oid.id_pub);
+	D_ASSERT(oca != NULL);
+	D_ASSERT(mrone->mo_iod_num <= DSS_ENUM_UNPACK_MAX_IODS);
+	for (i = 0; i < mrone->mo_iod_num; i++) {
+		D_ASSERT(mrone->mo_iods[i].iod_type == DAOS_IOD_SINGLE);
+
+		size = daos_iods_len(&mrone->mo_iods[i], 1);
+		D_ALLOC(data, size);
+		if (data == NULL)
+			D_GOTO(out, rc =-DER_NOMEM);
+
+		d_iov_set(&iov[i], data, size);
+		sgls[i].sg_nr = 1;
+		sgls[i].sg_nr_out = 1;
+		sgls[i].sg_iovs = &iov[i];
+	}
+
+	D_DEBUG(DB_REBUILD,
+		DF_UOID" mrone %p dkey "DF_KEY" nr %d eph "DF_U64"\n",
+		DP_UOID(mrone->mo_oid), mrone, DP_KEY(&mrone->mo_dkey),
+		mrone->mo_iod_num, mrone->mo_epoch);
+
+	rc = dsc_obj_fetch(oh, mrone->mo_epoch, &mrone->mo_dkey,
+			   mrone->mo_iod_num, mrone->mo_iods, sgls, NULL);
+	if (rc) {
+		D_ERROR("migrate dkey "DF_KEY" failed rc %d\n",
+			DP_KEY(&mrone->mo_dkey), rc);
+		D_GOTO(out, rc);
+	}
+
+	for (i = 0; i < mrone->mo_iod_num && DAOS_OC_IS_EC(oca); i++) {
+		daos_iod_t	*iod = &mrone->mo_iods[i];
+		uint32_t	start_shard;
+
+		start_shard = rounddown(mrone->mo_oid.id_shard,
+					obj_ec_tgt_nr(oca));
+		if (obj_ec_singv_one_tgt(iod, &sgls[i], oca)) {
+			D_DEBUG(DB_REBUILD, DF_UOID" one tgt.\n",
+				DP_UOID(mrone->mo_oid));
+			continue;
+		}
+
+		if (obj_shard_is_ec_parity(mrone->mo_oid, NULL)) {
+			d_iov_t e_iov = { 0 };
+
+			rc = obj_ec_singv_encode_buf(mrone->mo_oid.id_pub,
+						     iod, oca, &sgls[i],
+						     &e_iov);
+			if (rc)
+				return rc;
+			D_ASSERT(e_iov.iov_len <= sgls[i].sg_iovs[0].iov_len);
+			memcpy(sgls[i].sg_iovs[i].iov_buf, e_iov.iov_buf,
+			       e_iov.iov_len);
```

Comment:
here how to differentiate with parity target's parity data's address?
in above obj_ec_singv_encode_buf() returned e_iov, include all parity shards' data, need to take this parity shard's part right?

for example for 4+2 EC obj, when rebuild shard 5, but e_iov.iov_buf with parity data for both shard 4 and shard 5.

BTW, test degraded fetch after rebuild maybe can find some wrong.

Reply:
yes, this is something missing, I will deal with +2 in the following patch. This patch only deal with +1 at the moment.

Yes, the test will do degraded fetch to verify the data after rebuild the parity. see daos_rebuild_ec.c

Reply:
OK, fine to fix it in following patch.